indico/web/client/js/react/util/Responsive.jsx

```diff
@@ -0,0 +1,62 @@
+// This file is part of Indico.
+// Copyright (C) 2002 - 2019 CERN
+//
+// Indico is free software; you can redistribute it and/or
+// modify it under the terms of the MIT License; see the
+// LICENSE file for more details.
+
+import PropTypes from 'prop-types';
+import React from 'react';
+import Responsive from 'react-responsive';
+
+const DIMENSIONS = {
+  tablet: 768,
+  computer: 992,
+  wideScreen: 1200,
+};
+
+const factory = (minDimension, maxDimension) => {
+  /**
+   * This component extends `Responsive` from `react-responsive`, adding some
+   * useful configuration options.
+   */
+  function _SizeSpec({andLarger, andSmaller, orElse, children, onlyIf, ...restProps}) {
+    return (
+      <Responsive
+        minWidth={andSmaller ? null : minDimension}
+        maxWidth={andLarger ? null : maxDimension - 1}
+        {...restProps}
+      >
+        {matches => (matches && onlyIf ? children : orElse)}
+      </Responsive>
+    );
+  }
+
+  _SizeSpec.propTypes = {
+    /** this will make the rule match the specified screen size and above */
+    andLarger: PropTypes.bool,
+    /** this will make the rule match the specified screen size and below */
+    andSmaller: PropTypes.bool,
+    /** allows for negative cases to be specified easily */
+    orElse: PropTypes.node,
+    /** allows adding an extra condition to render the matching content */
+    onlyIf: PropTypes.bool,
+    children: PropTypes.node.isRequired,
+  };
+
+  _SizeSpec.defaultProps = {
+    andLarger: false,
+    andSmaller: false,
+    onlyIf: true,
+    orElse: null,
+  };
+
+  return _SizeSpec;
+};
+
+export default Object.assign(Responsive, {
```

Comment:
I've tried using the spread operator to clone the object like this:
```es6
export default {
  ...Responsive,
  WideScreen: factory(DIMENSIONS.wideScreen, null),
  Desktop: factory(DIMENSIONS.computer, DIMENSIONS.wideScreen),
  Tablet: factory(DIMENSIONS.tablet, DIMENSIONS.computer),
  Phone: factory(null, DIMENSIONS.tablet),
};
```
but then I get `Warning: React.createElement: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: object.`, any idea?

Reply:
`Responsive` is probably a react component so you can't just clone it like that (that's just copying object attrs). I guess in this case it's better to stick with the original option of modifying it - it's not pretty but less messy than creating a new class and inheriting from that (especially since that would break when they switch from class to function at some point)